### PR TITLE
TDL-17880 Add missing test cases

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change
- Updated the missing assertion in the discovery test
- Added the assertion of checking duplicate primary keys in pagination.
- Updated the assertion of duplicate primary keys in all_fields
- Added missing assertion in start_date test case
- Updated the create_object in utils.py

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/test_automatic_fields.py [hunk 1]
```diff
@@ -59,8 +57,10 @@ def test_run(self):
 
         # Run a sync job using orchestrator
         record_count_by_stream = self.run_and_verify_sync(conn_id)
+        synced_records = runner.get_records_from_target_output()
 
         actual_fields_by_stream = runner.examine_target_output_for_fields()
+        stream_primary_keys = self.expected_primary_keys()
 
         for stream in self.expected_streams().difference(untested_streams):
             with self.subTest(stream=stream):
```
Comment (namrata270998) [inline tests/test_automatic_fields.py:66]: We have skipped the primary keys assertion as for each streams in the sync mode, the tap first fetches the actual stream records and any updates via the events endpoint which also retrieves the created as well as updated records. Hence there will be 2 records of the same primary key.

File: tests/test_all_fields.py [hunk 2]
```diff
@@ -381,19 +395,26 @@ def all_fields_test(self, streams_to_test):
...
+                # Get the actual stream records based on the newly added field `updated_by_event_type` 
+                # as the events endpoints is not the latest version and hence returns deprecated fields also.
+                actual_record_message = actual_records.get('messages')
+                actual_records_data = [message['data'] for message in actual_record_message
+                                       if not message.get('data').get('updated_by_event_type', None)]
+
                 actual_records_keys = set()
                 for message in actual_record_message:
-                    if message['action'] == 'upsert':
+                    if message['action'] == 'upsert' and not message.get('data').get('updated_by_event_type', None):
```
Comment (dbshah1212) [inline tests/test_all_fields.py:406]: Add code comment to explain this check

File: tests/test_discovery.py [hunk 3]
```diff
@@ -74,6 +74,11 @@ def test_run(self):
                 self.assertTrue(len(stream_properties) == 1,
                                 msg="There is more than one top level breadcrumb")
 
+                actual_fields = [md_entry.get("breadcrumb")[1] for md_entry in metadata if md_entry.get("breadcrumb") != []]
+
+                # verify there are no duplicate metadata entries
+                self.assertEqual(len(actual_fields), len(set(actual_fields)), msg = f"duplicates in the fields retrieved")
```
Comment (savan-chovatiya) [inline tests/test_discovery.py:80]: Add a comment about this assertion at the function(test_run()) level comment above.

File: tests/test_pagination.py [hunk 4]
```diff
@@ -30,59 +30,64 @@ def test_run(self):
...
+
         # Select all streams and all fields within streams
         found_catalogs = self.run_and_verify_check_mode(conn_id)
-        our_catalogs = get_catalogs(conn_id, tested_streams)
+        our_catalogs = get_catalogs(conn_id, incremental_streams)
         self.select_all_streams_and_fields(conn_id, our_catalogs, select_all_fields=True)
 
         # Ensure tested streams have a record count which exceeds the API LIMIT
-        logging.info("Checking record counts for tested streams...")
+        print(" INFO Checking record counts for tested streams...")
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tests/test_pagination.py:60]: Using the logging, we were not able to log proper messages on the cci. Hence replaced with the print

File: tests/test_all_fields.py (hunk 2 above)
Comment (namrata270998) [reply tests/test_all_fields.py:407]: added comment

File: tests/test_discovery.py (hunk 3 above)
Comment (namrata270998) [reply tests/test_discovery.py:81]: Added comment

File: tests/test_start_date.py [hunk 5]
```diff
@@ -132,6 +131,21 @@ def test_run(self):
...
+                                       if message.get('action') == 'upsert']
+                primary_keys_list_2 = [tuple(message.get('data').get(expected_pk) for expected_pk in stream_primary_keys[stream])
+                                       for message in second_sync_records.get(stream).get('messages')
+                                       if message.get('action') == 'upsert']
+                primary_keys_sync_1 = set(primary_keys_list_1)
+                primary_keys_sync_2 = set(primary_keys_list_2)
+
+                # Verify by primary key values, that all records in the 1st sync are included in the 2nd sync.
+                self.assertTrue(
+                        primary_keys_sync_1.issubset(primary_keys_sync_2))
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tests/test_start_date.py:149]: Added comment

**Review by savan-chovatiya — APPROVED** [review]

File: tests/test_pagination.py [hunk 6]
```diff
@@ -30,59 +30,64 @@ def test_run(self):
...
             # 'charges',
             # 'coupons',
             # 'customers',
-            'disputes',
+            # 'disputes',
             # 'invoice_items',
+            # Data is generated automatically for 'invoice_line_items' when 'invoice_items' is created
             'invoice_line_items',
             # 'invoices',
             'payout_transactions',
```
Comment (dbshah1212) [inline tests/test_pagination.py:44]: Put comment after this that this is the substream and it's not possible to solve for substreams.

Comment (dbshah1212) [reply tests/test_pagination.py:44]: Also, attach the card number in the comment.

File: tests/utils.py [hunk 7]
```diff
@@ -472,12 +470,16 @@ def create_object(stream):
                 tax_rates=[],  # TODO tax rates
             )
 
-        if stream == 'charges':
+        if stream == 'charges' or stream == 'disputes':
```
Comment (dbshah1212) [inline tests/utils.py:473]: Why do we need to check differently for these 2 streams? Please explain in the comments.

**Review by dbshah1212 — APPROVED** [review]

**Review by savan-chovatiya — APPROVED** [review]

File: tests/utils.py (hunk 7 above)
Comment (namrata270998) [reply tests/utils.py:475]: Added in the comment.

**Comment by namrata270998** [comment]
This PR has the code form https://github.com/singer-io/tap-stripe/pull/123
Hence the PR #123 needs to be merged first

**Review by RushT007 — CHANGES_REQUESTED** [review]
@namrata270998:

- After execution of **test_all_fields.py** test, I observed that script is printing below error, 

`WARNING Stream[plans] Record[('plan_LHvXtOw4PlqR2c',)] Field[transform_usage] failed exact comparison.
AssertionError({'divide_by': 2, 'round': 'up'} != "{'divide_by': 2, 'round': 'up'}")`

There are such 49 warning messages observed during the execution. Since datatypes are different, equality assertion is always going to fail. Here approach should be comparing datatypes first then the values.

- Instead of 2 different log lines, can we print comprehensive single line exception here?

e.g. 

`WARNING:[DATATYPE MISMATCH] Stream[plans] Record[('plan_LHvXtOw4PlqR2c',)] Field[transform_usage]: AssertionError({'divide_by': 2, 'round': 'up'} != "{'divide_by': 2, 'round': 'up'}")`

`WARNING:[VALUE MISMATCH] Stream[plans] Record[('plan_LHvXtOw4PlqR2c',)] Field[transform_usage]: AssertionError({'divide_by': 2, 'round': 'up'} != "{'divide_by': 2, 'round': 'up'}")`

@kspeer825: 
In Stripe all fields execution, I observed that execution is producing almost 3K+ lines of warning messages. 

Need your help understanding below,

- What is the advantage of printing all those warning? 
- Are we going to take any action on those warning messages?
- Can we implement solution like just printing a diff between expectation and actual values?

Code under discussion,

`                                except AssertionError as failure_1:
                                    print(f"WARNING {base_err_msg} failed exact comparison.\n"
                                        f"AssertionError({failure_1})")
                                    nested_key = KNOWN_NESTED_MISSING_FIELDS.get(stream, {})
                                    if self.find_nested_key(nested_key, expected_field_value, field):
                                        continue
                                    if field in KNOWN_FAILING_FIELDS[stream] or field in FIELDS_TO_NOT_CHECK[stream]:
                                        continue # skip the following wokaround
                                    elif actual_field_value and field in FICKLE_FIELDS[stream]:
                                        self.assertIsInstance(actual_field_value, type(expected_field_value))
                                    elif actual_field_value:
                                        raise AssertionError(f"{base_err_msg} Unexpected field is being fickle.")
                                    else:
                                        print(f"WARNING {base_err_msg} failed datatype comparison. Field is None.")`

File: tests/test_all_fields.py [hunk 8]
```diff
@@ -37,33 +37,62 @@
     'invoices': set()
 }
 
+# we have observed that the SDK object creation returns some new fields intermittently
+SCHEMA_MISSING_FIELDS = {
+    'customers': {
+        'test_clock'
+    },
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tests/test_all_fields.py]: Updated the comment

**Review by KrisPersonal — APPROVED** [review]

**Comment by kspeer825** [comment]
@RushT007 
> What is the advantage of printing all those warning?
I think we originally just wanted to log as much information as possible 
> Are we going to take any action on those warning messages?
I think we hit a point of diminishing returns when testing this the first time around. So we left it in this state with the intention of refactoring it, when we had time to address it. I think the updates here are good, but we need to do an inventory of what test cases are actually covered here. I think the original pattern we used put us too far in the weeds. It took me about 20 minutes to figure out how this test was working and what it was covering. I think that's a problem. 
> Can we implement solution like just printing a diff between expectation and actual values? 
This can be done although I don't know if it would give us much value. I think this PR can be merged as the changes are good but we should rethink the approach and card out a refactor of this test.

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**